File: MovieCatalogueV2/favouritemovie/src/main/java/com/google/fdp/favouritemovie/MainActivity.java
package com.google.fdp.favouritemovie;

import android.content.Context;
import android.database.ContentObserver;
import android.database.Cursor;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.os.HandlerThread;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements LoadMoviesCallback {

    private List<Movie> movies = new ArrayList<>();
    private MovieAdapter movieAdapter;
    private DataObserver dataObserver;
    private Context mContext;

    RecyclerView rvMovie;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mContext = this;

        movieAdapter = new MovieAdapter(this, movies);
        movieAdapter.setOnItemClickCallback(new MovieAdapter.OnItemClickCallback() {
            @Override
            public void onItemClicked(Movie data) {
                Toast.makeText(mContext, "Clicked " + data.getTitle(), Toast.LENGTH_SHORT).show();
            }
        });

        rvMovie = findViewById(R.id.rvMovie);


        rvMovie.setAdapter(movieAdapter);
        rvMovie.setLayoutManager(new LinearLayoutManager(this));
        HandlerThread handlerThread = new HandlerThread("DataObserver");
        handlerThread.start();
        Handler handler = new Handler(handlerThread.getLooper());
        dataObserver = new DataObserver(handler, this);
        getContentResolver().registerContentObserver(DatabaseContract.MovieColumns.CONTENT_URI, true, dataObserver);
        new MovieAsyncTask(this, this).execute();
    }

    @Override
    public void postExecute(Cursor movieCursor) {
        if (movieCursor != null) {
            this.movies.clear();
            while (movieCursor.moveToNext()){
                this.movies.add(new Movie(movieCursor));
            }
            movieAdapter.notifyDataSetChanged();
        }
    }

    private static class MovieAsyncTask extends AsyncTask<Void, Void,Cursor> {
        private WeakReference<Context> weakContext;
        private WeakReference<LoadMoviesCallback> weakCallback;

        MovieAsyncTask(Context context, LoadMoviesCallback callback) {
            weakContext = new WeakReference<>(context);
            weakCallback = new WeakReference<>(callback);
        }

        @Override
        protected Cursor doInBackground(Void... voids) {
            if (weakContext.get() != null && weakContext.get().getContentResolver() != null) {
                return weakContext.get().getContentResolver().query(DatabaseContract.MovieColumns.CONTENT_URI, null, null, null, null);
            }

            return null;
        }

        @Override
        protected void onPostExecute(Cursor cursor) {
            super.onPostExecute(cursor);
            if (weakCallback.get() != null) {
                weakCallback.get().postExecute(cursor);
            }
        }
    }

    static class DataObserver extends ContentObserver {
        final Context context;

        DataObserver(Handler handler, Context context) {
            super(handler);
            this.context = context;
        }

        @Override
        public void onChange(boolean selfChange) {
            super.onChange(selfChange);
            new MovieAsyncTask(context, (MainActivity) context).execute();
        }
    }
}

File: MovieCatalog 3/favoritemovie/src/main/java/com/alfianh/favoritemovie/MainActivity.kt
package com.alfianh.favoritemovie

import android.content.Context
import android.database.ContentObserver
import android.database.Cursor
import android.os.AsyncTask
import android.os.Bundle
import android.os.Handler
import android.os.HandlerThread
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.alfianh.favoritemovie.DatabaseContract.MovieColumns.Companion.CONTENT_URI
import kotlinx.android.synthetic.main.activity_main.*
import java.lang.ref.WeakReference

class MainActivity : AppCompatActivity(), LoadMoviesCallback {

  private var movies: MutableList<Movie> = mutableListOf()
  private lateinit var movieAdapter: MovieAdapter
  private lateinit var dataObserver: DataObserver

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)
    movieAdapter = MovieAdapter(this, movies) {
      Toast.makeText(this, "Clicked ${movies[it].title}", Toast.LENGTH_SHORT).show()
    }
    rvMovie.adapter = movieAdapter
    rvMovie.layoutManager = LinearLayoutManager(this)
    val handlerThread = HandlerThread("DataObserver")
    handlerThread.start()
    val handler = Handler(handlerThread.looper)
    dataObserver = DataObserver(handler, this)
    contentResolver.registerContentObserver(CONTENT_URI, true, dataObserver)
    MovieAsyncTask(this, this).execute()
  }

  override fun postExecute(movieCursor: Cursor?) {
    movieCursor?.let {
      this.movies.clear()
      while (movieCursor.moveToNext()) {
        this.movies.add(Movie.fromCursor(movieCursor))
      }
      movieAdapter.notifyDataSetChanged()
    }
  }

  private class MovieAsyncTask constructor(context: Context, callback: LoadMoviesCallback) :
      AsyncTask<Void, Void, Cursor?>() {
    private var weakContext: WeakReference<Context> = WeakReference(context)
    private var weakCallback: WeakReference<LoadMoviesCallback> = WeakReference(callback)

    override fun doInBackground(vararg voids: Void): Cursor? {
      return weakContext.get()?.contentResolver?.query(CONTENT_URI, null, null, null, null)
    }

    override fun onPostExecute(data: Cursor?) {
      super.onPostExecute(data)
      weakCallback.get()?.postExecute(data)
    }
  }

  internal class DataObserver(handler: Handler, private val context: Context) :
      ContentObserver(handler) {
    override fun onChange(selfChange: Boolean) {
      super.onChange(selfChange)
      MovieAsyncTask(context, context as MainActivity).execute()
    }
  }
}
